File: Code_Jam/Tidy_Numbers/tidy_numbers.cpp
/*
A very nice problem that asks to find the last maximum "tidy" number(The digits from left to right be in ascending order),
smaller than the given number.

Author of this file:Lazaros Zervos
Use this code only for educational purposes.

Copyright@ 2018
*/



#include <iostream>
#include <string>
#include <vector>
#include <cstdbool>
#include <bits/stdc++.h>

using namespace std;





int main(void)	 {

  int No_Of_Trial;
  cin>>No_Of_Trial;

  for(int i=1; i<=No_Of_Trial; i++)  {
	long int tidy_number;

	long int number;	
	cin>>number;


	vector<long int> v;
	long int copy=number;
	while(copy/10 !=0) {
		int last_digit=copy%10;
		v.push_back(last_digit);
		copy=copy/10;
	}
	v.push_back(copy);
	reverse(v.begin(),v.end());

	int rem;
	bool flag=true;
	if(number/10==0){
	 cout<<"Case #"<<i<<": "<<number<<endl;
	 continue;
	}

	for(int z=0; z<v.size()-1; z++) {
		if(v[z]>v[z+1]) {
		  if(flag==false) {
			v[rem]-=1;
			for(int j=rem+1; j<v.size(); j++) {
					v[j]=9;
			}
			break;
		  }
		  else {
			v[z]-=1;
			for(int j=z+1; j<v.size(); j++) {
				v[j]=9;
			}
			break;
		  }
		}
		else if(v[z]==v[z+1] &&  flag==true) {
			rem=z;
			flag=false;
		}
		else if(v[z]<v[z+1]) {
		  flag=true;
		}
   }
   string s;
   for(int j=0; j<v.size(); j++){
	  s+=to_string(v[j]);
   }
   tidy_number=stol(s);
   cout<<"Case #"<<i<<": "<<tidy_number<<endl;
   v.clear();
 }
}

File: Code_Jam/Tidy_Numbers/tidy_numbers.py
def readCases(fileName):
	inputFile=open(fileName,'r')
	cases=inputFile.readline()
	cases=int(cases)
	return cases

def readNumbers(fileName,cases):
	inputFile=open(fileName,'r')
	number=inputFile.readline()
	for i in range(cases):	
		number=inputFile.readline()
	return int(number)

def main():
	cases=readCases('B-large-practice.in')
	for i in range(1,cases+1):
		number_list=[]
		s=" "
		number=readNumbers('B-large-practice.in',i)
		copy=number
		while copy//10!=0:
			last_digit=copy%10
			number_list.append(last_digit)
			copy=copy//10
		number_list.append(copy)
		number_list.reverse()

		flag=True
		if number//10 ==0:
			print("Case #{i}: {number}".format(i=i,number=number))
			continue

		for z in range(len(number_list)-1):
			if number_list[z]>number_list[z+1]:
				if flag==False:
					number_list[rem]-=1
					for j in range(rem+1,len(number_list)):
						number_list[j]=9
					break
				else:
					number_list[z]-=1
					for j in range(z+1,len(number_list)):
						number_list[j]=9
					break
			elif number_list[z]==number_list[z+1] and flag==True:
				rem=z
				flag=False
			elif number_list[z]<number_list[z+1]:
				flag=True

		for j in range(len(number_list)):
			s+=str(number_list[j])
		
		tidy_number=int(s)
		print("Case #{i}: {number}".format(i=i,number=tidy_number))
		number_list.clear()


if __name__ == '__main__':
	main()

